native/dormqr.go

```diff
@@ -0,0 +1,108 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/lapack"
+)
+
+// Dormqr multiplies the matrix c by the othogonal matrix q defined by the
+// slices a and tau. A and tau are as returned from Dgeqrf.
+//  C = Q * C    if side == blas.Left and trans == blas.NoTrans
+//  C = Q^T * C  if side == blas.Left and trans == blas.Trans
+//  C = C * Q    if side == blas.Right and trans == blas.NoTrans
+//  C = C * Q^T  if side == blas.Right and trans == blas.Trans
+// If side == blas.Left, a is a matrix of side k x m, and if side == blas.Right
+// a is of size k x n. This uses a blocked algorithm.
+//
+// work is a matrix for temporary memory. It must have size at m if Left and n if Right.
+// The length of work limits the available block size. As a special case, if lwork == -1,
+// work[0] will be populated with the optimal working length.
+//
+// Tau contains the householder scales and must have length at least k.
+func (impl Implementation) Dormqr(side blas.Side, trans blas.Transpose, m, n, k int, a []float64, lda int, tau, c []float64, ldc int, work []float64, lwork int) {
+	left := side == blas.Left
+	notran := trans == blas.NoTrans
+	if left {
+		checkMatrix(m, k, a, lda)
+	} else {
+		checkMatrix(n, k, a, lda)
+	}
+	checkMatrix(m, n, c, ldc)
+
+	const nbmax = 64
+	nw := n
+	if side == blas.Right {
+		nw = m
+	}
+	opts := string(side) + string(trans)
+	nb := min(nbmax, impl.Ilaenv(1, "DORMQR", opts, m, n, k, -1))
+	lworkopt := max(1, nw) * nb
+	if lwork == -1 {
+		work[0] = float64(lworkopt)
+		return
+	}
+	if left {
+		if lwork < n {
+			panic(badWork)
+		}
+	} else {
+		if lwork < m {
+			panic(badWork)
+		}
+	}
+	if m == 0 || n == 0 || k == 0 {
+		return
+	}
+	nbmin := 2
+
+	ldwork := nb
+	if nb > 1 && nb < k {
+		iws := nw * nb
+		if lwork < iws {
+			nb = lwork / nw
+			nbmin = max(2, impl.Ilaenv(2, "DORMQR", opts, m, n, k, -1))
+		}
+	}
+	if nb < nbmin || nb >= k {
+		// Call unblocked code
+		impl.Dorm2r(side, trans, m, n, k, a, lda, tau, c, ldc, work)
+		return
+	}
+	ldt := nb
+	t := make([]float64, nb*ldt)
+
+	switch {
```

Comment:
Alright, I changed the code around, but I really don't think it's simpler. The different cases loop in different directions in the different cases (top vs bottom, forward vs backward). This is disguised by looking common in the netlib code. It's easy to see in the separate case that, for example, n is fixed in left && notran, rather than having so see that ni is never modified.